File: services/payments/orchestrator.py
import time

from fastapi import HTTPException

from repositories.payment_repo import (
    create_payment_runtime,
    get_payment_bundle,
    get_payment_runtime_by_tx_ref,
    update_payment_runtime_status,
)
from schemas.payments_schema import (
    BundleType,
    CheckoutCreateRequest,
    CheckoutSessionOut,
    NormalizedWebhookEvent,
    PaymentProvider,
    PaymentRuntime,
    PaymentStatus,
)
from services.payments.fulfillment import fulfill_verified_payment
from services.payments.idempotency import mark_event_if_new
from services.payments.providers import FlutterwaveProvider, PaystackProvider, StripeProvider
from services.payments.routing import resolve_currency, resolve_provider


_provider_registry = {
    PaymentProvider.flutterwave: FlutterwaveProvider(),
    PaymentProvider.paystack: PaystackProvider(),
    PaymentProvider.stripe: StripeProvider(),
}


def _build_tx_ref(user_id: str, bundle_id: str) -> str:
    return f"uid:{user_id}|ts:{int(time.time())}|bid:{bundle_id}"


async def create_checkout(
    request: CheckoutCreateRequest,
    *,
    user_id: str,
    email: str,
    first_name: str | None,
    last_name: str | None,
) -> CheckoutSessionOut:
    bundle = await get_payment_bundle(bundle_id=request.bundleId)
    if bundle is None:
        raise HTTPException(status_code=404, detail="Payment bundle not found")
    if bundle.bundleType == BundleType.subscription_stars:
        raise HTTPException(status_code=400, detail="Use wallet endpoint for stars subscription purchase")
    if bundle.amount is None or bundle.amount <= 0:
        raise HTTPException(status_code=400, detail="Checkout requires a cash-priced bundle")

    provider = resolve_provider(request.countryCode, request.provider)
    currency = resolve_currency(request.countryCode)
    tx_ref = _build_tx_ref(user_id=user_id, bundle_id=request.bundleId)

    adapter = _provider_registry[provider]
    session = await adapter.create_checkout_session(
        request,
        amount=float(bundle.amount),
        currency=currency,
        tx_ref=tx_ref,
        email=email,
        first_name=first_name,
        last_name=last_name,
    )

    await create_payment_runtime(
        PaymentRuntime(
            txRef=tx_ref,
            userId=user_id,
            bundleId=request.bundleId,
            chapterId=request.chapterId,
            provider=provider,
            providerReference=session.providerReference,
            countryCode=request.countryCode.upper(),
            currency=currency,
            amount=float(bundle.amount),
            status=PaymentStatus.pending,
        )
    )

    return session


async def process_webhook(
    provider: PaymentProvider,
    *,
    raw_body: bytes,
    headers: dict[str, str],
) -> dict:
    adapter = _provider_registry[provider]
    event = await adapter.verify_webhook(raw_body=raw_body, headers=headers)

    is_new = await mark_event_if_new(
        provider=provider,
        event_id=event.eventId,
        tx_ref=event.txRef,
        provider_reference=event.providerReference,
    )
    if not is_new:
        return {"status": "idempotent_replay"}

    if not event.txRef:
        raise HTTPException(status_code=400, detail="Missing txRef in webhook event")

    runtime = await get_payment_runtime_by_tx_ref(tx_ref=event.txRef)
    if runtime is None:
        raise HTTPException(status_code=404, detail="Payment runtime not found")

    provider_ref = event.providerReference or runtime.providerReference
    if not provider_ref:
        raise HTTPException(status_code=400, detail="Missing provider reference")

    verification = await adapter.verify_transaction(provider_reference=provider_ref)
    if not verification.verified:
        await update_payment_runtime_status(tx_ref=runtime.txRef, status=PaymentStatus.failed)
        raise HTTPException(status_code=400, detail=verification.reason or "Verification failed")

    expected_amount = float(runtime.amount)
    if verification.amount is None or abs(float(verification.amount) - expected_amount) > 0.0001:
        await update_payment_runtime_status(tx_ref=runtime.txRef, status=PaymentStatus.failed)
        raise HTTPException(status_code=400, detail="Amount mismatch during verification")

    expected_currency = runtime.currency.upper()
    verified_currency = (verification.currency or "").upper()
    if not verified_currency or verified_currency != expected_currency:
        await update_payment_runtime_status(tx_ref=runtime.txRef, status=PaymentStatus.failed)
        raise HTTPException(status_code=400, detail="Currency mismatch during verification")

    await update_payment_runtime_status(
        tx_ref=runtime.txRef,
        status=PaymentStatus.verified,
        provider_reference=provider_ref,
    )

    fulfillment = await fulfill_verified_payment(
        user_id=runtime.userId,
        bundle_id=runtime.bundleId,
        tx_ref=runtime.txRef,
        chapter_id=runtime.chapterId,
    )

    await update_payment_runtime_status(
        tx_ref=runtime.txRef,
        status=PaymentStatus.fulfilled,
        provider_reference=provider_ref,
    )

    return {
        "status": "fulfilled",
        "txRef": runtime.txRef,
        "provider": provider.value,
        "fulfillment": fulfillment,
    }

File: mei-next/lib/payments/orchestrator.ts
/**
 * Cash-payment orchestrator — port of `services/payments/orchestrator.py`
 * (payments.md §3): initiate → webhook → idempotency → server-side re-verify
 * → amount/currency assert → fulfill.
 *
 * Status lifecycle on `payment_runtime`: initiated → pending → verified →
 * fulfilled (or → failed). Every transition writes `updatedAt = nowIso()`.
 */
import { db } from "@/lib/db";
import { envBool } from "@/lib/env";
import { HttpError } from "@/lib/http/errors";
import { PaymentRuntime, type PaymentRuntimeDoc, type UserDoc } from "@/lib/models";
import { nowEpoch, nowIso } from "@/lib/util/dates";
import { getPaymentBundle, normalizeBundleType } from "./bundles";
import type {
  CheckoutCreateRequest,
  CheckoutSessionOut,
  NormalizedWebhookEvent,
  PaymentProvider,
  PaymentProviderAdapter,
  PaymentStatus,
} from "./contracts";
import { fulfillVerifiedPayment } from "./fulfillment";
import { markEventIfNew } from "./idempotency";
import { flutterwaveProvider } from "./providers/flutterwave";
import { paystackProvider } from "./providers/paystack";
import { stripeProvider } from "./providers/stripe";
import { resolveCurrency, resolveProvider } from "./routing";

const providerRegistry: Record<PaymentProvider, PaymentProviderAdapter> = {
  flutterwave: flutterwaveProvider,
  paystack: paystackProvider,
  stripe: stripeProvider,
};

/** Ops kill-switch (improvements.md) — declared-but-unused in legacy, real here. */
function assertOrchestratorEnabled(): void {
  if (!envBool("PAYMENTS_ORCHESTRATOR_ENABLED", true)) {
    throw new HttpError(503, "Payments orchestrator is disabled");
  }
}

/** Lazy once-per-process runtime indexes (unique txRef + sparse provider/ref). */
let runtimeIndexesReady: Promise<unknown> | null = null;
function ensureRuntimeIndexes(): Promise<unknown> {
  if (!runtimeIndexesReady) {
    runtimeIndexesReady = PaymentRuntime.createIndexes().catch((err: unknown) => {
      runtimeIndexesReady = null;
      console.warn("payments: failed to ensure payment_runtime indexes", err);
    });
  }
  return runtimeIndexesReady;
}

function buildTxRef(userId: string, bundleId: string): string {
  return `uid:${userId}|ts:${nowEpoch()}|bid:${bundleId}`;
}

/** Status transition on payment_runtime — always bumps `updatedAt`. */
async function updateRuntimeStatus(
  txRef: string,
  status: PaymentStatus,
  providerReference?: string | null,
): Promise<void> {
  await db();
  const set: Record<string, unknown> = { status, updatedAt: nowIso() };
  if (providerReference) set.providerReference = providerReference;
  await PaymentRuntime.updateOne({ txRef }, { $set: set });
}

// ---------------------------------------------------------------------------
// 3A. Create checkout
// ---------------------------------------------------------------------------

/**
 * Port of `create_checkout`:
 * 1. bundle 404; reject subscriptionStars (wallet endpoint) and amount <= 0.
 * 2. resolve provider/currency by countryCode.
 * 3. txRef = "uid:{userId}|ts:{epochSeconds}|bid:{bundleId}".
 * 4. adapter checkout session; insert payment_runtime (status "pending").
 */
export async function createCheckout(user: UserDoc, body: CheckoutCreateRequest): Promise<CheckoutSessionOut> {
  assertOrchestratorEnabled();
  await db();

  const userId = String(user?._id ?? user?.userId ?? "");

  const bundle = await getPaymentBundle(body.bundleId);
  if (!bundle) throw new HttpError(404, "Payment bundle not found");
  if (normalizeBundleType(bundle.bundleType) === "subscriptionStars") {
    throw new HttpError(400, "Use wallet endpoint for stars subscription purchase");
  }
  if (bundle.amount === null || bundle.amount === undefined || Number(bundle.amount) <= 0) {
    throw new HttpError(400, "Checkout requires a cash-priced bundle");
  }

  const provider = resolveProvider(body.countryCode, body.provider ?? null);
  const currency = resolveCurrency(body.countryCode);
  const txRef = buildTxRef(userId, body.bundleId);
  const amount = Number(bundle.amount);

  const adapter = providerRegistry[provider];
  const session = await adapter.createCheckoutSession(body, {
    amount,
    currency,
    txRef,
    email: String(user?.email ?? ""),
    firstName: user?.firstName ?? null,
    lastName: user?.lastName ?? null,
  });

  await ensureRuntimeIndexes();
  const now = nowIso();
  await PaymentRuntime.create({
    txRef,
    userId,
    bundleId: body.bundleId,
    chapterId: body.chapterId ?? null,
    provider,
    providerReference: session.providerReference,
    countryCode: body.countryCode.toUpperCase(),
    currency,
    amount,
    status: "pending",
    createdAt: now,
    updatedAt: now,
  });

  return session;
}

// ---------------------------------------------------------------------------
// 3B. Process webhook
// ---------------------------------------------------------------------------

/**
 * Port of `process_webhook`:
 * 1. adapter signature verification on the RAW body.
 * 2. idempotency insert — duplicate → {status:"idempotent_replay"} and stop.
 * 3. txRef required (400); runtime by txRef (404); providerRef (400).
 * 4. server-side re-verify (never trust webhook amounts) — unverified →
 *    runtime "failed" + 400.
 * 5. amount tolerance 0.0001 and uppercase currency match — else "failed" + 400.
 * 6. runtime "verified" → fulfill → runtime "fulfilled".
 */
export async function processWebhook(
  provider: PaymentProvider,
  rawBody: string,
  headers: Record<string, string>,
): Promise<unknown> {
  assertOrchestratorEnabled();

  const adapter = providerRegistry[provider];
  const lowered: Record<string, string> = {};
  for (const [key, value] of Object.entries(headers)) lowered[key.toLowerCase()] = value;

  const event: NormalizedWebhookEvent = await adapter.verifyWebhook(rawBody, lowered);

  await db();
  const isNew = await markEventIfNew(
    provider,
    event.eventId,
    event.txRef ?? null,
    event.providerReference ?? null,
  );
  if (!isNew) return { status: "idempotent_replay" };

  if (!event.txRef) throw new HttpError(400, "Missing txRef in webhook event");

  await ensureRuntimeIndexes();
  const runtime = await PaymentRuntime.findOne({ txRef: event.txRef }).lean<PaymentRuntimeDoc>();
  if (!runtime) throw new HttpError(404, "Payment runtime not found");

  const providerRef = event.providerReference || runtime.providerReference;
  if (!providerRef) throw new HttpError(400, "Missing provider reference");
  const providerRefStr = String(providerRef);

  const verification = await adapter.verifyTransaction(providerRefStr);
  if (!verification.verified) {
    await updateRuntimeStatus(String(runtime.txRef), "failed");
    throw new HttpError(400, verification.reason || "Verification failed");
  }

  const expectedAmount = Number(runtime.amount);
  if (
    verification.amount === null ||
    verification.amount === undefined ||
    Math.abs(Number(verification.amount) - expectedAmount) > 0.0001
  ) {
    await updateRuntimeStatus(String(runtime.txRef), "failed");
    throw new HttpError(400, "Amount mismatch during verification");
  }

  const expectedCurrency = String(runtime.currency ?? "").toUpperCase();
  const verifiedCurrency = (verification.currency ?? "").toUpperCase();
  if (!verifiedCurrency || verifiedCurrency !== expectedCurrency) {
    await updateRuntimeStatus(String(runtime.txRef), "failed");
    throw new HttpError(400, "Currency mismatch during verification");
  }

  await updateRuntimeStatus(String(runtime.txRef), "verified", providerRefStr);

  const fulfillment = await fulfillVerifiedPayment({
    userId: String(runtime.userId),
    bundleId: String(runtime.bundleId),
    txRef: String(runtime.txRef),
    chapterId: runtime.chapterId ?? null,
  });

  await updateRuntimeStatus(String(runtime.txRef), "fulfilled", providerRefStr);

  return {
    status: "fulfilled",
    txRef: runtime.txRef,
    provider,
    fulfillment,
  };
}
